### PR TITLE
feat: Implement delta-based streaming and configurable streaming for AI chat models

### PR DESCRIPTION
**Pull request by adamwdraper**
### 🎯 Summary

This PR implements **delta-based streaming** and **configurable streaming** for `mo.ui.chat`, aligning marimo's streaming behavior with industry standards (OpenAI, Anthropic, AI SDK) while improving performance and flexibility.

**Key Changes:**
- ✅ **Delta-based streaming**: Models now yield individual chunks (deltas) instead of accumulated text, reducing bandwidth and improving responsiveness
- ✅ **Configurable streaming**: Added `stream` parameter (default: `True`) to all built-in chat models with automatic fallback for non-streaming models
- ✅ **Improved error handling**: Graceful fallback when streaming is not supported
- ✅ **Updated documentation and examples**: Clear guidance for custom model implementations

---

### 📊 What Changed

**9 files changed: +852 insertions, -233 deletions**

#### Core Changes
1. **`marimo/_plugins/ui/_impl/chat/chat.py`**: Updated to accumulate delta chunks from generators
2. **`marimo/_ai/llm/_impl.py`**: Refactored all 5 built-in models (OpenAI, Anthropic, Google, Groq, Bedrock) to:
   - Yield delta chunks instead of accumulated text
   - Support `stream=True/False` parameter
   - Include automatic fallback for non-streaming scenarios

#### Documentation & Examples
3. **`docs/api/inputs/chat.md`**: Added "How Streaming Works" section with delta vs accumulated comparison
4. **`examples/ai/chat/streaming_custom.py`**: Updated to demonstrate delta-based pattern
5. **`examples/ai/chat/README.md`**: Clarified streaming behavior
6. **`examples/ai/chat/streaming_openai.py`**: Removed (redundant after streaming became default)

#### Tests
7. **`tests/_plugins/test_chat_delta_streaming.py`**: New file with 10 comprehensive tests for delta streaming
8. **`tests/_ai/test_streaming_config.py`**: New file with 15 tests for configurable streaming
9. **`tests/_plugins/ui/_impl/chat/test_chat.py`**: Updated existing tests for delta-based streaming

---

#### Delta-Based Streaming
**Before:**
```python
def _stream_response(self, response):
    accumulated = ""
    for chunk in response:
        accumulated += chunk.content
        yield accumulated  # Send entire text every time
```

**After:**
```python
def _stream_response(self, response):
    for chunk in response:
        yield chunk.content  # Send only new content (delta)
```

**Benefits:**
- **Bandwidth**: Reduces data transfer by ~99% for long responses
- **Standards**: Matches OpenAI, Anthropic, and AI SDK patterns
- **Performance**: Frontend receives deltas and handles accumulation

#### Configurable Streaming
All models now support:
```python
# Enable streaming (default)
model = OpenAI("gpt-4", stream=True)

# Disable streaming (for models like o1-preview)
model = OpenAI("o1-preview", stream=False)
```

**Automatic Fallback:**
- If streaming fails with a model, automatically retries with `stream=False`
- Handles edge cases gracefully without user intervention

---

### 🧪 Testing

All tests pass ✅ (167 passed, 35 skipped):
- ✅ 10 tests for delta streaming logic (async/sync, edge cases, unicode)
- ✅ 15 tests for configurable streaming (all 5 models, fallback, error handling)
- ✅ 18 updated chat tests for delta-based streaming
- ✅ 124 existing AI tests still passing

Run tests:
```bash
hatch run +py=3.12 test:test tests/_ai/test_streaming_config.py
hatch run +py=3.12 test:test tests/_plugins/test_chat_delta_streaming.py
hatch run +py=3.12 test:test tests/_plugins/ui/_impl/chat/test_chat.py
```

---

#### For Users (Built-in Models)
```python
import marimo as mo

# Streaming enabled by default
chat = mo.ui.chat(
    mo.ai.llm.openai("gpt-4.1")
)

# Disable streaming for specific models
chat = mo.ui.chat(
    mo.ai.llm.openai("o1-preview", stream=False)
)
```

#### For Custom Models
```python
def custom_streaming_model(messages, config):
    """Custom model with delta-based streaming."""
    response = "Hello world from AI"
    for word in response.split():
        yield word + " "  # Yield deltas, not accumulated text
```

See updated docs: `docs/api/inputs/chat.md`

---

### 🔗 Related

This PR extracts the delta-based streaming improvements from #7258 **without** the LiteLLM migration, as per team feedback that LiteLLM cannot be a required dependency.

---

### ✅ Checklist

- [x] Core implementation completed
- [x] All tests passing (167 passed)
- [x] Documentation updated
- [x] Examples updated
- [x] No breaking changes (backward compatible with existing custom models)
- [x] No new required dependencies

---

> [!NOTE]
> Switches chat streaming to delta chunks across UI and built-in models with auto fallback when streaming isn’t supported, and updates docs, examples, and tests.
> 
> - **Streaming overhaul (delta-based)**:
>   - `marimo/_plugins/ui/_impl/chat/chat.py`: `_handle_streaming_response` now accumulates yielded deltas and emits incremental `stream_chunk` updates; returns full accumulated text.
>   - `marimo/_ai/llm/_impl.py`:
>     - Built-ins (`openai`, `anthropic`, `google`, `groq`, `bedrock`) now yield deltas (no accumulated text).
>     - Add `_looks_like_streaming_error` and implement try-stream-then-fallback to non-streaming for models that don’t support streaming.
> - **Docs & examples**:
>   - `docs/api/inputs/chat.md`: add “How Streaming Works” (delta vs accumulated) and custom model guidance.
>   - `examples/ai/chat/streaming_custom.py`: updated to delta-yielding pattern; `examples/ai/chat/README.md` clarifies default streaming behavior.
>   - Remove `examples/ai/chat/streaming_openai.py` (streaming now default).
> - **Tests**:
>   - New: `tests/_plugins/test_chat_delta_streaming.py`, `tests/_ai/test_streaming_config.py` (delta generation, fallback).
>   - Update: `tests/_plugins/ui/_impl/chat/test_chat.py`, `tests/_ai/llm/test_impl.py` to align with delta semantics.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 01b3ad9d6d61bc3be3ac12abc9311819bf8c1553. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: marimo/_ai/llm/_impl.py [hunk 1]
```diff
@@ -58,6 +58,8 @@ class openai(ChatModel):
             If not provided, the API key will be retrieved
             from the OPENAI_API_KEY environment variable or the user's config.
         base_url: The base URL to use
+        stream: Whether to stream responses. Defaults to True.
```
Comment (mscolnick) [inline marimo/_ai/llm/_impl.py]: we decided to leave this out and just always opt into streaming https://github.com/marimo-team/marimo/pull/7222
